File: projects/codecs/Matlab/src/tests/test_experiments/test_empty_experiment.m
function failures=test_empty_experiment()
        failures=0;
        totalTests=0;
        
        RL_init();
        

        for whichEpisode = [1:1:4]
            startTuple=RL_start();
            
            if mod(whichEpisode,2)==0
                [failures,totalTests]=check_fail(length(startTuple.a.intArray)~=0,failures,totalTests);
                [failures,totalTests]=check_fail(length(startTuple.a.doubleArray)~=0,failures,totalTests);
                [failures,totalTests]=check_fail(length(startTuple.a.charArray)~=0,failures,totalTests);

                [failures,totalTests]=check_fail(length(startTuple.o.intArray)~=0,failures,totalTests);
                [failures,totalTests]=check_fail(length(startTuple.o.doubleArray)~=0,failures,totalTests);
                [failures,totalTests]=check_fail(length(startTuple.o.charArray)~=0,failures,totalTests);
            else
                [failures,totalTests]=check_fail(length(startTuple.a.intArray)~=7,failures,totalTests);
                [failures,totalTests]=check_fail(length(startTuple.a.doubleArray)~=3,failures,totalTests);
                [failures,totalTests]=check_fail(length(startTuple.a.charArray)~=1,failures,totalTests);

                [failures,totalTests]=check_fail(length(startTuple.o.intArray)~=2,failures,totalTests);
                [failures,totalTests]=check_fail(length(startTuple.o.doubleArray)~=4,failures,totalTests);
                [failures,totalTests]=check_fail(length(startTuple.o.charArray)~=5,failures,totalTests);
            end
        for whichStep=0:1:4
    		stepTuple=RL_step();
            [failures,totalTests]=check_fail(stepTuple.terminal~=0,failures,totalTests);
            [failures,totalTests]=check_fail(stepTuple.terminal~=0,failures,totalTests);

            if mod(whichEpisode,2)==0
                [failures,totalTests]=check_fail(length(stepTuple.a.intArray)~=0,failures,totalTests);
                [failures,totalTests]=check_fail(length(stepTuple.a.doubleArray)~=0,failures,totalTests);
                [failures,totalTests]=check_fail(length(stepTuple.a.charArray)~=0,failures,totalTests);

                [failures,totalTests]=check_fail(length(stepTuple.o.intArray)~=0,failures,totalTests);
                [failures,totalTests]=check_fail(length(stepTuple.o.doubleArray)~=0,failures,totalTests);
                [failures,totalTests]=check_fail(length(stepTuple.o.charArray)~=0,failures,totalTests);
            else
                [failures,totalTests]=check_fail(length(stepTuple.a.intArray)~=7,failures,totalTests);
                [failures,totalTests]=check_fail(length(stepTuple.a.doubleArray)~=3,failures,totalTests);
                [failures,totalTests]=check_fail(length(stepTuple.a.charArray)~=1,failures,totalTests);

                [failures,totalTests]=check_fail(length(stepTuple.o.intArray)~=2,failures,totalTests);
                [failures,totalTests]=check_fail(length(stepTuple.o.doubleArray)~=4,failures,totalTests);
                [failures,totalTests]=check_fail(length(stepTuple.o.charArray)~=5,failures,totalTests);
            end
        end

	
	

        end

        
        
        
        
        
        

            
        if(failures>0)
            fprintf(1,'Failed %d of %d tests in test_empty_experiment\n',failures,totalTests);
        else
            fprintf(1,'Passed ALL  %d tests in test_empty_experiment\n',totalTests);
        end
        
        disconnectGlue();
end


function [failures, totalTests]=check_fail(theCondition, failures, totalTests)
    totalTests=totalTests+1;
    if(theCondition)
        fprintf(1,'Failed test %d\n',totalTests);
        failures=failures+1;
    end
end

File: tests/test_empty_experiment.c
/* 
* Copyright (C) 2007, Brian Tanner

 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at

     http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.

* 
*  $Revision$
*  $Date$
*  $Author$
*  $HeadURL$
* 
*/
#ifdef HAVE_CONFIG_H
#include <config.h>
#endif


#include <stdio.h>

#include <rlglue/RL_glue.h>
#include <string.h>

	int tests_failed=0;
	int test_count=0;

	void check_fail(int condition){
		test_count++;
		if(condition!=0){
			printf("Failed check: %d\n",test_count);
			tests_failed++;
		}
	}

	

int main(int argc, char *argv[]) {
	int whichEpisode=0;
	int whichStep=0;
	int i=0;
	observation_action_t startTuple={{0},{0}};
	reward_observation_action_terminal_t stepTuple={0};

	RL_init();
	
	for(whichEpisode=1;whichEpisode<5;whichEpisode++){
		startTuple=RL_start();
		
		if(whichEpisode%2==0){
			check_fail(startTuple.a.numInts!=0);
			check_fail(startTuple.a.numDoubles!=0);
			check_fail(startTuple.a.numChars!=0);

			check_fail(startTuple.o.numInts!=0);
			check_fail(startTuple.o.numDoubles!=0);
			check_fail(startTuple.o.numChars!=0);
		}else{
			
			check_fail(startTuple.a.numInts!=7);
			for(i=0;i<startTuple.a.numInts;i++)
				check_fail(startTuple.a.intArray[i]!=i);

			check_fail(startTuple.a.numDoubles!=3);
			for(i=0;i<startTuple.a.numDoubles;i++){
				check_fail(startTuple.a.doubleArray[i]-(double)i/(double)startTuple.a.numDoubles>.000001);
			}

			check_fail(startTuple.a.numChars!=1);
			for(i=0;i<startTuple.a.numChars;i++)
				check_fail(startTuple.a.charArray[i]!='a'+i);


			check_fail(startTuple.o.numInts!=2);
			for(i=0;i<startTuple.o.numInts;i++)
				check_fail(startTuple.o.intArray[i]!=i);

			check_fail(startTuple.o.numDoubles!=4);
			for(i=0;i<startTuple.o.numDoubles;i++){
				check_fail(startTuple.o.doubleArray[i]-(double)i/(double)startTuple.o.numDoubles>.000001);
			}

			check_fail(startTuple.o.numChars!=5);
			for(i=0;i<startTuple.o.numChars;i++){
				check_fail(startTuple.o.charArray[i]!='a'+i);
			}
		}
		
		for(whichStep=0;whichStep<5;whichStep++){
			stepTuple=RL_step();
			check_fail(stepTuple.terminal!=0);
			check_fail(stepTuple.r!=0);

			if(whichEpisode%2==0){
				check_fail(stepTuple.a.numInts!=0);
				check_fail(stepTuple.a.numDoubles!=0);
				check_fail(stepTuple.a.numChars!=0);

				check_fail(stepTuple.o.numInts!=0);
				check_fail(stepTuple.o.numDoubles!=0);
				check_fail(stepTuple.o.numChars!=0);
			}else{
				check_fail(stepTuple.a.numInts!=7);
				check_fail(stepTuple.a.numDoubles!=3);
				check_fail(stepTuple.a.numChars!=1);

				check_fail(stepTuple.o.numInts!=2);
				check_fail(stepTuple.o.numDoubles!=4);
				check_fail(stepTuple.o.numChars!=5);
			}
		}
	}
	

	if(tests_failed!=0)
		printf("Failed %d / %d checks in %s\n",tests_failed,test_count, __FILE__);
	else
		printf("Passed all %d checks in %s\n",test_count,__FILE__);
	return tests_failed;
}
